### ISSUE TITLE
linking monitormw site to hydroshare does not work

### ISSUE THREAD

**Issue by dbressler75** [ODM2/ODM2DataSharingPortal#475]
I get an error when I try link my site to hydroshare - it simply gives the attached error message
![error_linking_monmw_to_hydroshare](https://user-images.githubusercontent.com/39309124/77476752-8ab6c080-6df1-11ea-889c-77f636a41582.PNG)

**Comment by movingplaid** [comment]
> 
> 
> I get an error when I try link my site to hydroshare - it simply gives the attached error message
> ![error_linking_monmw_to_hydroshare](https://user-images.githubusercontent.com/39309124/77476752-8ab6c080-6df1-11ea-889c-77f636a41582.PNG)

Just to confirm, I get this error as well when clicking the link from the [Account page](https://monitormywatershed.org/account/) when logged into the MonitorMW site. 

Unsure what action is intended afterwards, so I am not much help besides the obvious.

**Comment by SRGDamia1** [comment]
This is still not functional.

**Comment by dbressler75** [comment]
Yeah I've had this issue too with Hydroshare.  Sure wish there was a way we
could get these bugs worked on.  I'm going to keep bringing it up.

Sara, have you chatted with Anthony about it recently by chance?

David W. Bressler
Stroud Water Research Center
410-456-1071 (mobile)
610-268-2153 x312 (office)
dbressler@stroudcenter.org

**Comment by aufdenkampe** [comment]
From Martin Seul via a Sept 3, 2020 email "RE: integration of HydroShare with MonitorMyWatershed":

I reached out to Jeff and he kindly provided some more detail on the implementation and pointed to the authentication as potential problem that might have broken the App. I checked into this and the authentication requirements for apps using Hydroshare has changed to address a vulnerability concern, It requires a minor but essential changes, I have provided an example below I got from Tony Castronova”
 
Yes, we had a similar problem with some of our web applications. It resulted from a security update to the Oauth2 library to prevent passing client_secret in the HTTP header. The solution is to put all OAuth credentials in the message body. Below are a couple examples:

### Python3

```py
http_client = AsyncHTTPClient()  
# POST request parameters for HydroShare    
params = dict(grant_type='authorization_code',
              code=code,               
              client_id=self.client_id,
              client_secret=self.client_secret,               
              redirect_uri=self.oauth_callback_url)
              
url = url_concat( 'hydroshare.org/o/token/', params).split('?')   
req = HTTPRequest(url[0], method="POST", body=url[1], headers={"Accept": "application/json"}, validate_cert=False,)   
resp = yield http_client.fetch(req)
```

### Curl

```sh
#!/bin/bash
 
YOUR_CLIENT_ID=''
YOUR_CLIENT_SECRET=''
CALLBACK='127.0.0.1:80';
CODE=$1 # retrieved from hydroshare.org/o/authorize?
 
curl -k --header "Accept:: application/json" \
--request POST \
--data "grant_type=authorization_code&code=$CODE&client_id=$YOUR_CLIENT_ID&client_secret=$YOUR_CLIENT_SECRET&redirect_uri=$CALLBACK" \   
hydroshare.org/o/token
```
 
This could be the explanation for the problem.
I don’t have a setup to test this but this could be a starting point.
 
Happy to hop on a call to discuss.
